### DISCUSSION TITLE
Switching to NATS for worker-engine or engine-engine communication

### DISCUSSION THREAD

**Discussion by abelanger5** [hatchet-dev/hatchet#224] [Ideas]
The goal of Hatchet is to provide a scalable and distributed system for managing tasks. Eventually we would like to support geographic distribution of workers that connect to a set of brokers living close to the workers.

In Hatchet architecture terms, we would like the dispatchers to live close to the workers and write results to a nearby data store. We are still a long way from working out the consistency and replication model here. 

The first version of Hatchet is using RabbitMQ to broker messages between different engine components. Many users and people giving feedback have suggested switching to [NATS](https://nats.io/) instead. I like this idea particularly for engine-engine communication, because it will be easier to isolate and scale our engine components via nats than amqp. 

I am uncertain about worker-engine communication, because we get a lot out of the box with gRPC. Starting this thread to entertain some thoughts and ideas from the community and add my own input when we start mapping it out in more detail.

**Comment by geekodour** [comment]
I think it'll be nice to have NATS both for engine-engine and worker-engine just for consistency. 

Referring to https://github.com/hatchet-dev/hatchet/issues/89, do we have a choice to go NATS or postgres for this now? or I am misunderstanding something?

**Reply by yvesaston** [reply]
Indeed, looks like now we are not forced to use RabbitMQ to broker messages, and instead use a Postgre DB for the matter. Less dependency, great !

**Comment by yvesaston** [comment]
@abelanger5, likewise the folks you mention, I indeed see that using NATS in Hatchet would be great, for all communication. Related to your question about replacing gRPC by NATS for worker-engine communication ; for sure gRPC is a premium choice there ; you might though want to take a look at [nRPC](https://github.com/nats-rpc/nrpc).

I have also a question ; if you were to use NATS in Hatchet, would it also be a good idea to then use it as well for persistence ? NATS Jetstream comes with both a KV Store and an Object Store, potentially enough to be the persistence layer instead of a DB (less dependencies !).

**Comment by Inquerito** [comment]
If I'm not mistaken, antfu applied this previously

**Reply by yvesaston** [reply]
Hi, you mean it has been tried ? Where can I see the results ?

**Comment by kinthaiofficial** [comment]
NATS vs gRPC/Redis for worker-engine communication is a real architectural decision with different tradeoffs at agent workload scale.

NATS strengths for agent workloads:
- **Built-in pub/sub for fan-out** — when one agent event needs to notify multiple interested parties (parent agent, audit logger, budget tracker), NATS subjects handle this natively without additional infrastructure
- **At-most-once + at-least-once flexibility** — core NATS for fire-and-forget status events; JetStream for durable task queues that survive engine restarts
- **Lightweight** — lower overhead than Kafka for the medium message volumes typical in agent coordination (10K-100K msgs/sec is plenty)

For agent systems specifically:
- **Request-reply patterns** — NATS has native request-reply that's useful for capability negotiation ("worker, can you handle task type X?") without the overhead of HTTP round-trips
- **Subject-based routing** — `agent.{agent_id}.task.{task_id}.cost` as a subject pattern lets you subscribe to exactly the cost events you care about for a specific agent, without filtering
- **Budget events** — when an agent crosses 80% budget utilization, publish to `agent.{id}.budget.warning`. The engine subscribes and can pause the agent without polling

The challenge with NATS for agent coordination is ordering guarantees — if the sequence of "task started → delegated → completed" must be processed in order, JetStream with sequence numbers is necessary.

We're using a similar event bus pattern for KinthAI's agent coordination layer: https://blog.kinthai.ai/221-agents-multi-agent-coordination-lessons covers the coordination design.

Is the NATS migration primarily for performance, or for the pub/sub patterns that are awkward over gRPC?
